docs/tutorial/part_1.rst

```diff
@@ -0,0 +1,238 @@
+
+    from .models import Room, Message, User
+    from .serializers import MessageSerializer, RoomSerializer, UserSerializer
+
+    class RoomConsumer(ObserverModelInstanceMixin, GenericAsyncAPIConsumer):
+        queryset = Room.objects.all()
+        serializer_class = RoomSerializer
+        lookup_field = "pk"
+
+        async def disconnect(self, code):
+            if hasattr(self, "room_subscribe"):
+                await self.remove_user_from_room(self.room_subscribe)
+                await self.notify_users()
+            await super().disconnect(code)
+
+        @action()
+        async def join_room(self, pk, **kwargs):
+            self.room_subscribe = pk
+            await self.add_user_to_room(pk)
+            await self.notify_users()
+
+        @action()
+        async def leave_room(self, pk, **kwargs):
+            await self.remove_user_from_room(pk)
+
+        @action()
+        async def create_message(self, message, **kwargs):
+            room:Room = await self.get_room(pk=self.room_subscribe)
+            await database_sync_to_async(Message.objects.create)(
+                room=room, 
+                user=self.scope["user"],
+                text=message
+            )
+
+        @action()
+        async def subscribe_to_messages_in_room(self, pk, **kwargs):
+            await self.message_activity.subscribe(room=pk)
+
+        @model_observer(Message)
+        async def message_activity(self, message, observer=None, **kwargs):
+            await self.send_json(message)
+
+        @message_activity.groups_for_signal
+        def message_activity(self, instance: Message, **kwargs):
+            yield 'room__%s' % instance.room_id
+            yield 'pk__%s' % instance.pk
+
+        @message_activity.groups_for_consumer
+        def message_activity(self, room=None, **kwargs):
+            if room is not None:
+                yield 'room__%s' % room
+
+        @message_activity.serializer
+        def message_activiy(self, instance:Message, action, **kwargs):
+            return dict(data=MessageSerializer(instance).data, action=action.value, pk=instance.pk)
+
+        async def notify_users(self):
+            room:Room = await self.get_room(self.room_subscribe)
+            for group in self.groups:
+                await self.channel_layer.group_send(
+                    group,
+                    {
+                        'type':'update_users',
+                        'usuarios':await self.current_users(room)
+                    }
+                )
+
+        async def update_users(self, event:dict):
+            await self.send(text_data=json.dumps({'usuarios':event["usuarios"]}))
+    
+        @database_sync_to_async
+        def get_room(self, pk:int)->Room:
+            return Room.objects.get(pk=pk)
+
+        @database_sync_to_async
+        def current_users(self, room:Room):
+            return [UserSerializer(user).data for user in room.current_users.all()]
```

Comment:
What about doing `UserSerializer(room.current_users.all(), many=True).data`?

Reply:
That is really stupid on my part! jaja you are right!